### ISSUE TITLE
Implement text range selection in TextView

### ISSUE THREAD

**Issue by uxmal** [uxmal/reko#30]
Recall that the RichEdit Control was unsuitable for reko, and Windows Forms has no other way of displaying rich text. TextView is a poor man's version of GtkTextView which seems to serve adequately. The problem is that there is no way for users to mark a _text selection_. This precludes copying snippets of code, which is a fairly reasonable expectation. Today reko cheeses out by providing a "Copy All" verb, but it certainly could be better.

To implement this, TextView needs the usual assortment of Selection range plumbing: where the selection starts (the anchor), where it ends. A TextSpan must be able to perform hit-detection based on an X-coordinate and return the index position within the span that was selected. TextView needs a CurrentSelection that returns a TextSelection object consisting of all the TextSpans, both partially and fully selected. TextSpans, when rendering, must know what portion of them is selected and what is not so that they render colors correctly.

**Comment by nemerle** [comment]
Yup, I think we'll need to spice that control up pretty heavily in the future:
1. adding tooltips on registers/instructions
2. visualizing additional data ( points where something was 'assumed' by the user etc. )
3. Identifying various bit's and pieces of the current shown data

When i was doing my boomerang/exetoc assembly view experiments i did it like this:
* for each visible assembly instruction/label - produce a simple tree-like structure, embedding various parts in proper tags ( <offset> <label> <instruction>, <register>, <label>, ... ) 
* Render the pre-processed tree into colored Qt rich text  document and using it's services to identify elements under mouse cursor, so i could show relevant tooltips, context menu's etc.

**Comment by uxmal** [comment]
thats the same approach I'm taking with reko. The drawback is I don't want to add a dependecy on Qt or Gtk. It strikes me, however, that adding tooltips for registers and instructions shouldn't be _that_ bad, it's just that there is no selection visible. Let me mull on this for a while

I'll create a separate issue to track the notion of annotations in the TextView.

**Comment by nemerle** [comment]
Agreed, no need to pull in gtk or qt 
As for selection handling, one way of doing it would be to disable selection handling on control itself but add mouse handling by-hand ( handling clicks and mouse moves ), and mark areas as selected programmaticly, although i guess it would be a bit of PITA :smile:

**Comment by uxmal** [comment]
Better to make TextView text-selection aware. However, all annotatable things, like registers, instructions, and identifiers, can indicate their annotatedness very easily. The class `Reko.Gui.Windows.DisassemblyFormatter` could tell `TextView` about being annotatable.

**Comment by nemerle** [comment]
Yes, I can see that having a proper per-character selection would be helpful when getting assembly out of reko. 
I think we'll still need some kind of 
```
List<object> GetObjectsFromCurrentSelection();
```
that would return the 'root' objects in current selection

**Comment by uxmal** [comment]
Each `TextSpan`  the TextView can be associated with a `Tag`: an arbitrary object. This tag could be what gets returned from `GetSelectedObjects()`

**Comment by uxmal** [comment]
I just checked into `master` the new TextView which allows arbitrary selection of _text_. Basically, I've implemented a simple rich text editor.... (what this has to do with decompiling is an exercise left to the reader!) 

Next step is to provide an API to select the objects embedded in the selected text, so that relevant tool tips and context menus can be displayed.

**Comment by xor2003** [comment]
Maybe you can use [http://www.scintilla.org/Steps.html](Scintilla) as edit control. It is used for example in Notepad++

**Comment by uxmal** [comment]
@xor2003: thanks for the suggestion! However, a fullblown text editor with undo, at this time, is a bit overkill for Reko's needs.
